Post-mortem June 20 (#415)
openSUSE/obs-landing: _posts/deployments/2023-06-20-post-mortem.md

---
layout: post

> title: "Post-mortem: deletion of an architecture (i586)"

Reviewer: ```suggestion
title: "Post-mortem: Staging Workflow Unavailable"
```

category: deployments
---

> In efforts to add new repository Debian 12, architecture i586 was deleted.

Reviewer: ```suggestion
In our efforts to add a distribution (Debian 12), we accidentally deleted an entry in our architecture database table. That made all attempts to fetch this architecture through associations crash.
```

## Impact

> [Staging dashboard](https://build.opensuse.org/staging_workflows/openSUSE:Factory) for [openSUSE:Factory](https://build.opensuse.org/staging_workflows/openSUSE:Factory) was not accessible.

Reviewer: ```suggestion
The [staging workflow for openSUSE:Factory](https://build.opensuse.org/staging_workflows/openSUSE:Factory) was not accessible for 34 minutes.
```

## Root Causes

> A distribution could have more than one architectures. We added a new repository for Debian 12 with two architectures .i.e., i586 and x86_64. But after reading release post from Debian we realise i586 processors are not supported. While removing the unsupported architecture, i586 got deleted from every distribution.

Reviewer: That is not exactly what happened. We did not delete `Architecture` from all `Distribution`. We deleted the `Architecture` itself. It's used in a lot of places, the staging workflow for instance has nothing to do with `Distribution`.

Author: True. I've addressed it

## Trigger

> OBS was not able to find i586 architecture for any repository.

Reviewer: ```suggestion
Loading (through associations) the `Architecture` we deleted from the database.
```

## Detection

> We started receiving exceptions in Errbit error catcher.

Reviewer: ```suggestion
Our [people on call](https://trello.com/c/IKZgDuaD) started to see exceptions pile up in our [exception tracker](https://errbit.opensuse.org/).
```

## Resolution

> We created a new architecture entry for i586 in database.

Reviewer: ```suggestion
We re-created the row in the `Architecture` table in our database.
```

## Action items

> - Discuss better solutions to perform such actions.

Reviewer: This is not an Action Item. Action items are tracked somewhere, like an issue or card. 

Also better solutions than what? You do not mention what we did anywhere...

Author: I think action items don't apply here. There are follow-up tasks to address it or anything to track.

Reviewer: I don't think so no. Here are two action items I can think of:

- Developer documentation about the `Distribution` feature
  - what is it?
  - how to setup a distribution from an already existing project?
  - how to setup a distribution from scratch?

- Admin interface to:
  - create a `Distribution` from an existing project
  - create a `Distribution` from scratch

## Lessons learned

> Do not perform admin/management tasks for data deleteion in production until you are sure because it inherent risks of using it incorrectly and breaking things.

Reviewer 2: ```suggestion
Do not perform admin/management tasks for data deletion in production until you are sure because it inherent risks of using it incorrectly and breaking things.
```

Reviewer: This is not a lesson learned. We knew this before. It just has gone wrong. What's the real lesson here? :)

Author: It's updated. Please have a look

## What Went Well?

> - Once the issue was analyzed, we were able to fully fixed the issue in couple of minutes.

Reviewer 2: ```suggestion
- Once the issue was analyzed, we were able to fully fix the issue in couple of minutes.
```

Reviewer: This is true for every issue right? Once you analyzed it it's quick to fix. I think what you are trying to say is more something for where we got lucky. See below...

> - We were able to find the id of deleted record in matter of seconds.

Reviewer: I think this is where we got lucky. If we would not have this in terminal history somewhere....

## What Went Wrong?

- It took us 30 minutes to know about the issue after it's the first occurence.

> - The errors detected on Errbit were not clear and we were not able to realise the issue at first glance.

Reviewer: Not sure what you want to say with this? What was not clear?

Author: Exceptions were clear but we didn't realize soon enough about the frequency of those exceptions.

## Where We Got Lucky

> Information we needed to create new entry in database was readily available.

Reviewer: How was it readily available?

## Timeline (times in UTC)

> - 20/06/2023 12:43 - Detected some errors on errbit and got informed that the Factory Dashboard returns 500 errors.

Reviewer: ```suggestion
- 20/06/2023 12:43 - Our on call people detected some errors on errbit
- 20/06/2023 12:50 - Our on call people checked with customers and they confirmed the problem with the staging workflow
```

> - 20/06/2023 13:09 - Find the root cause.

Reviewer: How? Where? What took 20 minutes?

Author: It's updated as well. Please check

> - 20/06/2023 13:17 - Implemented solution of the problem.

Reviewer: ```suggestion
- 20/06/2023 13:16 - Re-created the Architecture in the database
- 20/06/2023 13:16 - Verified that the exceptions stopped
```

---
layout: post
title: "Post-mortem: Staging Workflow Unavailable"
category: deployments
---

In our efforts to add a distribution (Debian 12), we accidentally deleted an entry in our architecture database table. That made all attempts to fetch this architecture through associations crash.

## Impact

Various pages from the [staging workflow for openSUSE:Factory](https://build.opensuse.org/staging_workflows/openSUSE:Factory) were not accessible for 34 minutes.

## Root Causes

> A distribution could have more than one architectures. We added a new repository for Debian 12 with two architectures .i.e., i586 and x86_64. But after reading release post from Debian we realise i586 processors are not supported. While removing the unsupported architecture, i586 got deleted from architectures database table.

Reviewer: ```suggestion
A distribution can have an association to more than one architecture. We added a distribution for Debian 12 with two architectures: i586 and x86_64. But after reading [the release post from Debian](https://www.debian.org/News/2023/20230610) more closely we realized, i586 processors are not supported anymore. Then we accidentally deleted the unsupported architecture from the database table instead of deleting the association.
```

> A distribution could have more than one architectures. We added a new repository for Debian 12 with two architectures .i.e., i586 and x86_64. But after reading release post from Debian we realize i586 processors are not supported. While removing the unsupported architecture, i586 got deleted from architectures database table.

Reviewer: ```suggestion
A `Distribution` can have an association to more than one `Architecture`. We added a new `Distribution` for Debian 12 with two architectures: i586 and x86_64. But after reading [the release post from Debian](https://www.debian.org/News/2023/20230610) more closely, we realized that i586 processors are not supported anymore. While removing the unsupported architecture, we accidentally deleted it from the architectures database table instead of deleting the association.
```

## Trigger

Loading (through associations) the `Architecture` we deleted from the database.

## Detection

Our [people on call](https://trello.com/c/IKZgDuaD) started to see exceptions pile up in our [exception tracker](https://errbit.opensuse.org/).

## Resolution

We re-created the row in the `Architecture` table in our database.

## Action items

- Developer documentation about the Distribution feature answering questions like, what is it, how to setup a distribution from an already existing project, and how to setup a distribution from scratch.

> - An admin interface to create a Distribution from an existing project, and create a Distribution from scratch.

Reviewer: We have a documentation and a code bug tracker you can create tickets in...

Reviewer: So you are not going to create and link them? :)

Reviewer: A freestyle card or feature issue about creating an admin interface?

Author: Done. I've created an issue in our freestyle and it is linked here. Same for the docu and I'll take care of writing developer docu.

## Lessons learned

Do not perform admin/management tasks for data deletion in production until you try these operations in staging or development environments. It's even better to have a second pair of eyes with you. Mistakes can happen but the chances are low when people are collectively working on something.

## What Went Well?

- We were able to find the id of deleted record in matter of seconds though terminal history.

## What Went Wrong?

- It took us 30 minutes to know about the issue after it's the first occurrence.
- We did not realize about the frequency of exceptions and doubted that the exceptions are not related to the operations we performed recently.

## Where We Got Lucky

- Information we needed to create new entry in database was readily available in our shell history so we did not have to pull them out of some backup.
- We used a method to [delete](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-delete) the Architecture that did not trigger any callbacks or changes to the associations of the object destroyed. As this would have had way more impact.

> - We relatively quickly noticed exceptions as this did not throw an alert.

Reviewer: ```suggestion
- We relatively quickly noticed exceptions as this did not throw an alert.
- During the time the Architecture was deleted only customers using the staging workflow code path tried to fetch it. Other code-paths would have thrown exceptions too but nobody used them....
```

## Timeline (times in UTC)

> - 20/06/2023 12:43 - Our on call people detected some errors on Errbit.
> - 20/06/2023 12:50 - Our on call people checked with customers and they confirmed the problem with the staging workflow.
> - 20/06/2023 13:09 - Started a call to distribute [incident management roles](https://github.com/openSUSE/open-build-service/wiki/Incident-Management-Protocol) and find the root cause that the query we ran actually removed the architecture row from architectures table.
> - 20/06/2023 13:16 - Re-created the Architecture in the database.
> - 20/06/2023 13:16 - Verified that the exceptions stopped.

Reviewer: ```suggestion
- 20/06/2023 12:32 - We accidentally deleted the architecture
- 20/06/2023 12:43 - Our on call people detected exceptions piling up
- 20/06/2023 12:50 - Our on call people checked with customers and they confirmed a problem with the staging workflow
- 20/06/2023 12:52 - Our on call people started to alert the team
- 20/06/2023 13:03 - We discovered our mistake that caused this
- 20/06/2023 13:07 - Our on call people declared the incident according to our  [incident management protocol](https://github.com/openSUSE/open-build-service/wiki/Incident-Management-Protocol) 
- 20/06/2023 13:16 - Re-created the Architecture in the database.
- 20/06/2023 13:16 - Verified that the exceptions stopped.
- 20/06/2023 13:17 - Our on call people declared the incident resolved
```

## Root Causes

A `Distribution` can have an association to more than one `Architecture`. We added a new `Distribution` for Debian 12 with two architectures: i586 and x86_64. But after reading [the release post from Debian](https://www.debian.org/News/2023/20230610) more closely, we realized that i586 processors are not supported anymore. While removing the unsupported architecture, we accidentally deleted it from the architectures database table instead of deleting the association.
## Trigger

Loading (through associations) the `Architecture` we deleted from the database.

## Detection

Our [people on call](https://trello.com/c/IKZgDuaD) started to see exceptions pile up in our [exception tracker](https://errbit.opensuse.org/).

## Resolution

We re-created the row in the `Architecture` table in our database.

## Action items

- Developer documentation about the Distribution feature answering questions like, what is it, how to setup a distribution from an already existing project, and how to setup a distribution from scratch.
- An admin interface to create a Distribution from an existing project, and create a Distribution from scratch.
- Introduce incident management drills.

> - Add Grafana alert for 500 internal server errors.

Reviewer: ```suggestion
- Improve alerting for requests that result in internal server errors
```

## Trigger

Loading (through associations) the `Architecture` we deleted from the database.

## Detection

Our [people on call](https://trello.com/c/IKZgDuaD) started to see exceptions pile up in our [exception tracker](https://errbit.opensuse.org/).

## Resolution

We re-created the row in the `Architecture` table in our database.

## Action items

- [Developer documentation](https://github.com/openSUSE/open-build-service/wiki/Distributions) about the Distribution feature answering questions like, what is it, how to setup a distribution from an already existing project, and how to setup a distribution from scratch.
- An [admin interface](https://trello.com/c/K8a7nBZC/129-creating-a-distribution-admin-interface) to create a Distribution from an existing project, and create a Distribution from scratch.

> - Introduce incident management drills.

Reviewer: ```suggestion
- [Introduce incident management drills](https://trello.com/c/GijvIK2Z).
```

> - Improve [alerting](https://trello.com/c/MEGZQSLh/130-grafana-alerts-for-requests-that-result-in-internal-server-errors) for requests that result in internal server errors.

Reviewer: ```suggestion
- Improve [alerting](https://trello.com/c/MEGZQSLh) for requests that result in internal server errors.
```
